### PR TITLE
OSOE-425: Automatically add Jira issue key prefix and link to pull requests in Lombiq.GitHub.Actions

### LINKED ISSUE #86: Automatically add Jira issue key prefix and link to pull requests (OSOE-425)
- In the `validate-pull-request` workflow, implement automation to prefix a newly opened PR's title with the corresponding Jira issue key, as well as add a link to the issue into the PR's description, if these are not there already. The issue key can be determined from the branch name.
- Don't do anything if the branch name doesn't match an issue key pattern (like for community contributions).
- In all OSOCE submodules, change `verify-submodule-pull-request` workflows to `validate-pull-request`. These should include a call to a new, `validate-submodule-pull-request` reusable workflow, that runs both the submodule verification and the issue key insertion. The `validate-this-pull-request` workflow should be adapated.

[Jira issue](https://lombiq.atlassian.net/browse/OSOE-425)

### PR DESCRIPTION
**Pull request by Luko6**
[OSOE-425](https://lombiq.atlassian.net/browse/OSOE-425)

Fixes #86

### CODE REVIEW

File: .github/actions/prefix-pr/index.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+const core = require("@actions/core");
```
Comment (Piedone) [inline .github/actions/prefix-pr/index.js]: Don't use JS but PS: https://lombiq.atlassian.net/wiki/spaces/DEV/pages/2545614855/Writing+GitHub+Actions+workflows+and+actions#:~:text=Prefer%20to%20use%20PowerShell%207%2B%20scripts.

File: .github/actions/prefix-pr/action.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+name: "Prefix PR"
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: You also have to do something with the `create-jira-issues-for-community-activities` workflow because otherwise it'll create issues for every PR, see Teams.

Comment (Luko6) [reply .github/actions/prefix-pr/action.yml]: So if I understand that workflow correctly, it first checks if the title contains an issue key or not, and if it doesn't it creates a Jira issue. How about checking the branch instead of the title there, and determining if its an issue branch or not, instead of the PR title?

Comment (Piedone) [reply .github/actions/prefix-pr/action.yml]: Issue-specific PRs are possible to do from non-issue branches too (like https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/pull/211) so it needs to keep checking the title.

Comment (Luko6) [reply .github/actions/prefix-pr/action.yml]: How about checking for both and going forward if one is true?

Comment (Piedone) [reply .github/actions/prefix-pr/action.yml]: That sounds good. DRY the branch-checking logic with the prefixing action; see the existing Scripts folder and how we load scripts from there.

Comment (Luko6) [reply .github/actions/prefix-pr/action.yml]: Between these two there is nothing to really DRY.

File: .github/actions/verify-submodule-pull-request/Check-Parent.ps1 [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
     [string] $Repository,
-    [string] $Title
+    [string] $Branch
 )
 
 $url = "https://api.github.com/repos/$Repository/pulls?state=open&per_page=100"
 $titles = curl -s -H 'Accept: application/vnd.github.v3+json' $url | ConvertFrom-Json | % { $_.title }
 
-$issueCode = $Title -replace '^\s*(\w+-\d+)\s*:.*$', '$1'
+$issueCode = $Branch.replace("issue/", "")
```
Comment (Piedone) [inline .github/actions/verify-submodule-pull-request/Check-Parent.ps1]: Note that branches can also have names like `issue/OSOE-123-some-side-quest`. Same for _Prefix-PR.ps1_.

File: .github/workflows/validate-pull-request.yml [hunk 4]
```diff
@@ -7,7 +7,13 @@ jobs:
   validate-pull-request:
     runs-on: ubuntu-latest
     steps:
+    - name: Prefix PR
+      id: prefix
```
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml]: Why is this needed? In the other workflows too.

File: .github/workflows/validate-pull-request.yml [hunk 5]
```diff
@@ -7,7 +7,13 @@ jobs:
   validate-pull-request:
     runs-on: ubuntu-latest
     steps:
+    - name: Prefix PR
```
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml]: Use a name that conveys what kind of prefixing happens. Also in all the other places.

File: .gitignore [hunk 6]
```diff
@@ -9,3 +9,4 @@ node_modules/
 *.user
 .pnpm-debug.log
 /.editorconfig
+!**/prefix-pr/node_modules
```
Comment (Piedone) [inline .gitignore]: This isn't needed.

File: Docs/Workflows.md [hunk 7]
```diff
@@ -129,19 +129,19 @@ When `source` is not provided, it assumes a default value of pushing to the [Lom
 
 Valid values for `verbosity` are those defined by [MSBuild](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference?view=vs-2022#:~:text=you%20can%20specify%20the%20following%20verbosity%20levels). The default value is `minimal`.
 
-## Submodule verify workflow
+## Submodule validate workflow
```
Comment (Piedone) [inline Docs/Workflows.md:133]: Change the wording below too.

File: .github/workflows/prefix-pr.yml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+name: "Prefix PR"
```
Comment (Piedone) [inline .github/workflows/prefix-pr.yml]: This workflow isn't needed.

File: .github/actions/prefix-pr/Prefix-PR.ps1 [hunk 9]
```diff
@@ -0,0 +1,46 @@
+param(
+    [string] $Github_Repository,
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR.ps1]: Similarly the others:

```suggestion
    [string] $GitBubRepository,
```

File: .github/actions/prefix-pr/Prefix-PR.ps1 [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    [string] $Github_Ref,
+    [string] $Branch,
+    [string] $Github_Token
+)
+
+$jiraBaseUrl = "https://lombiq.atlassian.net/browse/"
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR.ps1]: This should rather come from a parameter of the action. Supply the `DEFAULT_JIRA_BASE_URL` secret to it in our workflows, which is "https://lombiq.atlassian.net" (see how this is used by `create-jira-issues-for-community-activities`).

File: .github/actions/prefix-pr/Prefix-PR.ps1 [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+$owner, $repo = $Github_Repository.Split('/')
+$prId = $Github_Ref -replace "refs\/pull\/(\d+)\/merge", '$1'
+
+$url = "https://api.github.com/repos/$owner/$repo/pulls/$prId"
+$headers = @{"Authorization" = "Bearer $Github_Token"; "Accept" = "application/vnd.github+json"}
+
+$pr = Invoke-WebRequest $url -Headers $headers | % {$_.Content} | ConvertFrom-Json
+$title = $pr.title
+$body = $pr.body
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR.ps1]: These are available in the [`github.event.pull_request` context](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request) already. `$prId` is contained there too.

File: .github/actions/prefix-pr/Prefix-PR.ps1 [hunk 12]
```diff
@@ -0,0 +1,46 @@
+param(
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR.ps1]: Name the file in a way tha conveys what kind of prefixing happens.

File: .github/actions/prefix-pr/action.yml [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+runs:
+  using: "composite"
+  steps:
+    - name: Setup Scripts
+      shell: pwsh
+      run: |
+        "${{ github.action_path }}" >> $Env:GITHUB_PATH
+        (Resolve-Path "${{ github.action_path }}/../../../Scripts").Path >> $Env:GITHUB_PATH
+
+    - name: Prefix PR
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: This needs fork check like `check-merge-conflict`.

**Review by Piedone — CHANGES_REQUESTED** [review]
In the PR description, add a "fixes" reference to https://github.com/Lombiq/GitHub-Actions/issues/86.

File: .github/actions/prefix-pr/action.yml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+name: "Prefix PR"
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: Add a description, then a line break before `inputs`.

**Comment by Piedone** [comment]
Add a fixes reference to https://github.com/Lombiq/GitHub-Actions/issues/86 to the PR description.

File: .github/workflows/validate-pull-request.yml (hunk 4 above)
Comment (Luko6) [reply .github/workflows/validate-pull-request.yml]: It's not needed, was needed with an earlier iteration.

File: .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1 [hunk 15]
```diff
@@ -0,0 +1,45 @@
+param(
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1]: Leverage Git's rename next time please, because now this file appears as brand new, making review harder than it should be.

File: .github/actions/verify-submodule-pull-request/Check-Parent.ps1 [hunk 16]
```diff
@@ -6,7 +6,8 @@ param(
 $url = "https://api.github.com/repos/$Repository/pulls?state=open&per_page=100"
 $titles = curl -s -H 'Accept: application/vnd.github.v3+json' $url | ConvertFrom-Json | ForEach-Object { $_.title }
 
-$issueCode = $Branch.replace("issue/", "")
+$Branch -match '(\w+-\d+)'
+$issueCode = $matches[0]
```
Comment (Piedone) [inline .github/actions/verify-submodule-pull-request/Check-Parent.ps1:10]: Where does `$matches` come from?

File: .github/workflows/validate-pull-request.yml [hunk 17]
```diff
@@ -7,13 +7,15 @@ jobs:
...
+      env:
+        JIRA_BASE_URL: ${{ secrets.JIRA_BASE_URL }}
       with:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
     - name: Check for Merge Conflict in PR
       uses: Lombiq/GitHub-Actions/.github/actions/check-merge-conflict@issue/OSOE-425
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        JIRA_BASE_URL: ${{ secrets.JIRA_BASE_URL }}
```
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml]: That action doesn't (and should) use the Jira base URL.

File: Docs/Workflows.md [hunk 18]
```diff
@@ -132,7 +132,7 @@ Valid values for `verbosity` are those defined by [MSBuild](https://docs.microso
 
 ## Submodule validate workflow
 
-Verifies if the submodule contains a JIRA style issue code (e.g. PROJ-123) and if a pull request exists for the parent module. Example _publish.yml_:
```
Comment (Piedone) [inline Docs/Workflows.md:135]: Does it only validated this?

File: .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1 [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    [string] $JiraBaseUrl,
+    [string] $GitBubRepository,
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1]:
```suggestion
    [string] $GitHubRepository,
```

File: .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1 [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    [string] $GithubRef,
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1]:
```suggestion
    [string] $GitHubRef,
```

File: .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1 [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    [string] $Branch,
+    [string] $GithubToken,
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1]:
```suggestion
    [string] $GitHubToken,
```

File: .github/workflows/validate-pull-request.yml [hunk 22]
```diff
@@ -7,13 +7,15 @@ jobs:
   validate-pull-request:
     runs-on: ubuntu-latest
     steps:
-    - name: Prefix PR
-      id: prefix
+    - name: Prefix PR with issue key
```
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml]: Prefix PR how, what are you prefixing? (Also, "prefix" might not be a good word to use here if you don't want a very long name.) Also, follow the casing pattern of other names. Same for the action name/display name.

File: .github/workflows/validate-pull-request.yml [hunk 23]
```diff
@@ -7,13 +7,15 @@ jobs:
...
       uses: Lombiq/GitHub-Actions/.github/actions/prefix-pr@issue/OSOE-425
+      env:
+        JIRA_BASE_URL: ${{ secrets.JIRA_BASE_URL }}
```
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml:21]: This isn't declared as a secret.

File: .github/workflows/validate-pull-request.yml (hunk 22 above)
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml]: Also test this workflow in the test repo.

File: .github/actions/prefix-pr/action.yml [hunk 24]
```diff
@@ -1,4 +1,6 @@
-name: "Prefix PR"
+name: "Prefix PR with issue key"
+description: "Prefixes the PR title with a Jira issue key, and also adding the key to the body if not present"
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]:
```suggestion
description: "Prefixes the PR title with a Jira issue key, and also adds the key to the body if not present."
```

File: .github/actions/prefix-pr/action.yml [hunk 25]
```diff
@@ -1,4 +1,6 @@
...
+
 inputs:
   GITHUB_TOKEN:
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: image

But this shouldn't be an input anyway, but a secret.

File: .github/actions/prefix-pr/action.yml [hunk 26]
```diff
@@ -12,6 +14,10 @@ runs:
...
 
-    - name: Prefix PR
+    - name: Prefix PR with issue key
+      # If this is run for a PR opened from a fork (not from a branch in the given repo) then "Resource not accessible
+      # by integration" errors can happen. Thus, not running it for forks during the pull_request event (but it can
+      # run during push, for example). See https://github.com/Lombiq/GitHub-Actions/issues/72 for details.
+      if: github.event.pull_request == '' || github.event.pull_request.head.repo.fork == false
       shell: pwsh
-      run: Prefix-PR $Env:GITHUB_REPOSITORY $Env:GITHUB_REF $Env:GITHUB_HEAD_REF ${{ inputs.GITHUB_TOKEN }}
+      run: Prefix-PR-with-issue-key ${{ env.JIRA_BASE_URL }} $Env:GITHUB_REPOSITORY $Env:GITHUB_REF $Env:GITHUB_HEAD_REF ${{ inputs.GITHUB_TOKEN }} "${{ github.event.pull_request.title }}" "${{ github.event.pull_request.body }}" ${{ github.event.pull_request.number }}
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: Line too long. Use PowerShell splatting.

File: .github/actions/prefix-pr/action.yml [hunk 27]
```diff
@@ -12,6 +14,10 @@ runs:
         "${{ github.action_path }}" >> $Env:GITHUB_PATH
         (Resolve-Path "${{ github.action_path }}/../../../Scripts").Path >> $Env:GITHUB_PATH
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: Are the common scripts actually needed?

File: .github/actions/prefix-pr/action.yml (hunk 26 above)
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: Both `env` and `$Env` are used. Decide on one.

File: .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1 [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+    [string] $Title,
+    [string] $Body,
+    [string] $prId
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1]: Follow naming conventions.

File: .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1 (hunk 15 above)
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1]: Follow naming conventions with the file's name.

File: .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1 [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+)
+
+$jiraBaseUrl = $JiraBaseUrl + '/browse/';
+$owner, $repo = $GitBubRepository.Split('/')
```
Comment (Piedone) [inline .github/actions/prefix-pr/Prefix-PR-with-issue-key.ps1]: Nice, I didn't know you can deconstruct in PS.

File: .github/actions/prefix-pr/action.yml (hunk 24 above)
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: No need to quote these.

**Review by Piedone — CHANGES_REQUESTED** [review]

File: .github/actions/verify-submodule-pull-request/Check-Parent.ps1 (hunk 16 above)
Comment (Luko6) [reply .github/actions/verify-submodule-pull-request/Check-Parent.ps1:10]: It's an automatic variable coming from `-match` ([read this](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_regular_expressions?view=powershell-7.3#groups-captures-and-substitutions) if interested)

File: Docs/Workflows.md (hunk 18 above)
Comment (Luko6) [reply Docs/Workflows.md:135]: Extended this.

File: .github/workflows/validate-pull-request.yml [hunk 30]
```diff
@@ -2,12 +2,25 @@ name: Validate Pull Request
 
 on:
   workflow_call:
+    secrets:
+      # We can't access org secrets here so they need to be passed in.
+      JIRA_BASE_URL:
+        required: true
```
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml]: Make this non-required, and use our Jira base URL as the default. That way, it's backward compatible.

Comment (Luko6) [reply .github/workflows/validate-pull-request.yml]: Can I leave the submodule calls as they are? Or should the passing of the secret be deleted?

File: .github/actions/verify-submodule-pull-request/Check-Parent.ps1 (hunk 16 above)
Comment (Piedone) [reply .github/actions/verify-submodule-pull-request/Check-Parent.ps1:10]: Ah OK, didn't know that, thanks.

File: .github/workflows/validate-pull-request.yml (hunk 30 above)
Comment (Piedone) [reply .github/workflows/validate-pull-request.yml]: Please delete them.

File: .github/actions/prefix-pr/action.yml [hunk 31]
```diff
@@ -1,9 +1,5 @@
-name: "Prefix PR with issue key"
-description: "Prefixes the PR title with a Jira issue key, and also adding the key to the body if not present"
-
-inputs:
-  GITHUB_TOKEN:
-    required: true
+name: Validate Jira Issue Code in Pull Request
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: The action still has the old name, as does the script.

File: .github/actions/prefix-pr/action.yml [hunk 32]
```diff
@@ -12,12 +8,21 @@ runs:
...
+        $params = @{
+          JiraBaseUrl = $Env:JIRA_BASE_URL ? $Env:JIRA_BASE_URL : "https://lombiq.atlassian.net";
+          GitHubRepository = $Env:GITHUB_REPOSITORY;
+          Branch = $Env:GITHUB_HEAD_REF;
+          GitHubToken = $Env:GITHUB_TOKEN;
+          Title = "${{ github.event.pull_request.title }}";
+          Body = "${{ github.event.pull_request.body }}";
+          PullRequestID = ${{ github.event.number }};
+        }
+        PrefixPRWithIssueKey @params
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: Don't use abbreviations like PR in identifiers.

File: .github/workflows/validate-submodule-pull-request.yml [hunk 33]
```diff
@@ -19,14 +18,15 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
-    - name: Prefix PR with issue key
+    - name: Validate Jira Issue Code in Pull Request
```
Comment (Piedone) [inline .github/workflows/validate-submodule-pull-request.yml]:
```suggestion
    - name: Add Jira Issue Code to Pull Request
```
?

It doesn't actually do any validation. It won't fail if the issue code is not there.

File: .github/actions/prefix-pr/action.yml [hunk 34]
```diff
@@ -1,9 +1,5 @@
-name: "Prefix PR with issue key"
-description: "Prefixes the PR title with a Jira issue key, and also adding the key to the body if not present"
-
-inputs:
-  GITHUB_TOKEN:
-    required: true
+name: Validate Jira Issue Code in Pull Request
+description: Prefixes the PR title with a Jira issue key, and also adds the key to the body if not present
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: As I [pointed this before](https://github.com/Lombiq/GitHub-Actions/pull/90#discussion_r1028417820):

```suggestion
description: Prefixes the PR title with a Jira issue key, and also adds the key to the body if not present.
```

File: .github/workflows/validate-pull-request.yml [hunk 35]
```diff
@@ -2,20 +2,25 @@ name: Validate Pull Request
 
 on:
   workflow_call:
+    secrets:
+      # We can't access org secrets here so they need to be passed in.
+      JIRA_BASE_URL:
```
Comment (Piedone) [inline .github/workflows/validate-pull-request.yml:7]: image

File: .github/actions/prefix-pr/action.yml [hunk 36]
```diff
@@ -12,12 +8,21 @@ runs:
...
+    - name: Validate Jira Issue Code in Pull Request
       # If this is run for a PR opened from a fork (not from a branch in the given repo) then "Resource not accessible
       # by integration" errors can happen. Thus, not running it for forks during the pull_request event (but it can
       # run during push, for example). See https://github.com/Lombiq/GitHub-Actions/issues/72 for details.
       if: github.event.pull_request == '' || github.event.pull_request.head.repo.fork == false
       shell: pwsh
-      run: Prefix-PR-with-issue-key ${{ env.JIRA_BASE_URL }} $Env:GITHUB_REPOSITORY $Env:GITHUB_REF $Env:GITHUB_HEAD_REF ${{ inputs.GITHUB_TOKEN }} "${{ github.event.pull_request.title }}" "${{ github.event.pull_request.body }}" ${{ github.event.pull_request.number }}
+      run: |
+        $params = @{
+          JiraBaseUrl = $Env:JIRA_BASE_URL ? $Env:JIRA_BASE_URL : "https://lombiq.atlassian.net";
```
Comment (Piedone) [inline .github/actions/prefix-pr/action.yml]: Trailing semicolons aren't needed.

**Review by Piedone — CHANGES_REQUESTED** [review]

**Review by Piedone — APPROVED** [review]

File: .github/actions/verify-submodule-pull-request/Check-Parent.ps1 [hunk 37]
```diff
@@ -1,12 +1,13 @@
...
-    [string] $Title
+    [string] $Branch
 )
 
 $url = "https://api.github.com/repos/$Repository/pulls?state=open&per_page=100"
 $titles = curl -s -H 'Accept: application/vnd.github.v3+json' $url | ConvertFrom-Json | ForEach-Object { $_.title }
 
-$issueCode = $Title -replace '^\s*(\w+-\d+)\s*:.*$', '$1'
+$Branch -match '(\w+-\d+)'
+$issueCode = $matches[0]
```
Comment (0liver) [inline .github/actions/verify-submodule-pull-request/Check-Parent.ps1:10]: This failed for me here: https://github.com/Lombiq/Orchard-Chart.js/pull/47/checks. Any idea why?

Comment (Luko6) [reply .github/actions/verify-submodule-pull-request/Check-Parent.ps1:10]: Is it a standardly formatted issue branch?

Comment (0liver) [reply .github/actions/verify-submodule-pull-request/Check-Parent.ps1:10]: Click the link, go, try!

image
